### ISSUE TITLE
Allow volumetric and/or surface area based spawning of particles with (skinned) mesh renderers

### ISSUE THREAD

**Issue by Banane9** [Yellow-Dog-Man/Resonite-Issues#3247]
### Is your feature request related to a problem? Please describe.

> Correct me if I'm wrong, but in the video it looks like the skinned mesh emitter is based on the vertex density of the mesh, i.e. where there's more vertices, there's more particles. Would it be at all possible to allow this to be  surface area or volume based as well? That would greatly improve the effect for many applications, imo. For example in the video, it's kinda distracting from the effect, that the particles are much denser around the head and hands than the body when moving, imo. 

 _Originally posted by @Banane9 in [#587](https://github.com/Yellow-Dog-Man/Resonite-Issues/issues/587#issuecomment-2504175696)_

Referring to the video on skinned mesh renderers as particle emitters:
https://youtu.be/CCrP-bi02H0

### Describe the solution you'd like

Allow volumetric and/or surface area based spawning of particles with a (skinned) mesh renderer to produce an equal density of particles in the mesh's volume.

### Describe alternatives you've considered

All other modes are biased towards areas of higher mesh density.

### Requesters

[Bb]anane9

**Comment by Redd56** [comment]
this reminds me of something i need to test and may add an issue for. mostly for froox to test in the new system

**Comment by Frooxius** [comment]
This is something I plan on adding, but it's a bigger chunk of work. Main thing we need is implement collection types that allow sampling of random elements with probability equivalent to the size of the element.

**Comment by Banane9** [comment]
Glad to hear!

I can imagine it's not the easiest thing to do, especially with the _variable_ nature of skinned mesh surfaces / volumes.

**Comment by Frooxius** [comment]
For SkinnedMeshes it wouldn't take into account current deformations - that would require recalculating this acceleration structure every frame, which would get *really* heavy.

**Comment by Banane9** [comment]
Yea, that's true - rest pose is perfectly fine too.

**Comment by Redd56** [comment]
> For SkinnedMeshes it wouldn't take into account current deformations - that would require recalculating this acceleration structure every frame, which would get _really_ heavy.


Would timeslicing the calculations maybe make it less terrible for performance 
Alao I wouldn't expect but static blendshapes to be evaluated. Maybe once a second blendshapes are evaluated at most. Super time slicing

**Comment by Frooxius** [comment]
No. That would be computed on a background thread.

The problem is that it'll introduce latency for relatively minor effect in most cases and it'll eat a lot CPU usage.

**Comment by Redd56** [comment]
Tho i would like to see one that is a snapshot of the current deformations. so things like the effect of engineer tf2's teleporter could be properly made (tho that requires xomponent access for dynamiclt getting it) example: (at 1:54 its explained) https://youtu.be/TQINJNbQlb0

**Comment by TisFoolish** [comment]
> For SkinnedMeshes it wouldn't take into account current deformations - that would require recalculating this acceleration structure every frame, which would get really heavy.

What about a secondary, extremely stripped down acceleration structure that is *just* the area of the triangles and positions of the verts, all calculated in a compute shader? One that is used just for the triangle based area particle spawning.

**Comment by Frooxius** [comment]
@TisFoolish Thats what the acceleration structure already is - it's just for the areas of the triangles to fetch them efficiently.

Compute Shader isn't super suitable here, it's not a workload that fits that level of parallelizarion.

The whole premise is kinda bad here. It's like asking for SkinnedMeshCollider - it's theoretically possible, but really inefficient and will use lots of cycles for things that can have much better alternate solutions.

**Comment by TisFoolish** [comment]
> Compute Shader isn't super suitable here, it's not a workload that fits that level of parallelizarion. 

I completely disagree there. When talking about avatars that is the high tens to hundreds of thousands of triangles that'd each need their area calculated. The math is simple, but the amount of tris a lot of avatars users have easily puts this in the realm of being quite suitable for a compute shader

**Comment by Frooxius** [comment]
@TisFoolish That's not sufficient though. The big part of this isn't calculating the areas, but actually generating the acceleration structure, which is typically a tree based acceleration structure, which GPU's aren't very good at building and it's typically computed on the CPU side.

Going part on GPU and part on CPU also isn't guaranteed to make things faster, because GPU <---> CPU transfers are relatively inefficient and have higher latency (you need to schedule computations to happen on the GPU).

I'm not really super interested in spending time debating this further. This is very niche and unless there's significant interest in skinned meshes specifically, there's no point in spending time on "HOW", when the question "SHOULD WE" wasn't answered first.

If you really want skinned meshes with realtime updates weights to be a thing, I recommend making separate issue and getting a bunch of upvotes for it for us to consider it, but for purposes of this issue we'll be doing static meshes only.

**Comment by Frooxius** [comment]
This has been added in 2025.12.29.1301 now!
